src/Workspaces/Remote/ServiceHub/Services/ProjectTelemetry/RemoteProjectTelemetryIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ProjectTelemetry;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal class RemoteProjectTelemetryIncrementalAnalyzer : IncrementalAnalyzerBase
+    {
+        /// <summary>
+        /// Channel back to VS to inform it of the designer attributes we discover.
+        /// </summary>
+        private readonly RemoteEndPoint _endPoint;
+
+        private readonly object _gate = new object();
+        private readonly Dictionary<ProjectId, ProjectTelemetryInfo> _projectToInfo = new Dictionary<ProjectId, ProjectTelemetryInfo>();
+
+        public RemoteProjectTelemetryIncrementalAnalyzer(RemoteEndPoint endPoint)
+            => _endPoint = endPoint;
+
+        /// <summary>
+        /// Collects data from <paramref name="project"/> and reports it to the telemetry service.
+        /// </summary>
+        /// <remarks>
+        /// Only sends data to the telemetry service when one of the collected data points changes, 
+        /// not necessarily every time this code is called.
+        /// </remarks>
+        public override async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
+        {
+            if (!semanticsChanged)
+                return;
+
+            var projectId = project.Id;
+            var language = project.Language;
+            var analyzerReferencesCount = project.AnalyzerReferences.Count;
+            var projectReferencesCount = project.AllProjectReferences.Count;
+            var metadataReferencesCount = project.MetadataReferences.Count;
+            var documentsCount = project.DocumentIds.Count;
+            var additionalDocumentsCount = project.AdditionalDocumentIds.Count;
+
+            var info = new ProjectTelemetryInfo
+            {
+                ProjectId = projectId,
+                Language = language,
+                AnalyzerReferencesCount = analyzerReferencesCount,
+                ProjectReferencesCount = projectReferencesCount,
+                MetadataReferencesCount = metadataReferencesCount,
+                DocumentsCount = documentsCount,
+                AdditionalDocumentsCount = additionalDocumentsCount,
+            };
+
+            lock (_gate)
+            {
+                if (_projectToInfo.TryGetValue(projectId, out var existingInfo) &&
```

Comment:
Should this also report if the project version is later than the last reported?  I don't know if this can happen, but if AnalyzeProjectAsync was invoked concurrently for the same project, then it could end up reporting info for p before p'.  p' would be dropped by the telemetry reporter since p comes after p' in the reporting queue.

Reply:
so that can't happen.  it's an invariant of hte incremetnal analyzer.  tagging @mavasani  just for extra paranoia

Reply:
Yep, Cyrus is correct - the work items for analysis callbacks are all queued so they cannot execute out of order.